blocksync/buffer.go

```diff
@@ -30,89 +30,57 @@ type blockBuffer struct {
 	bc              blockchain.Blockchain
 	ap              actpool.ActPool
 	size            uint64
-	startHeight     uint64
 	confirmedHeight uint64
 }
 
 // Flush tries to put given block into buffer and flush buffer into blockchain.
 func (b *blockBuffer) Flush(blk *blockchain.Block) (bool, bCheckinResult) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
-	l := logger.With().Uint64("recvHeight", blk.Height()).Uint64("startHeight", b.startHeight).Uint64("confirmedHeight", b.confirmedHeight).Str("source", "blockBuffer").Logger()
-
-	var (
-		syncedHeight uint64
-		moved        bool
-	)
-
+	moved := false
 	// check
 	h := blk.Height()
 	if h <= b.confirmedHeight {
 		return moved, bCheckinLower
 	}
-	if h < b.startHeight {
-		if err := commitBlock(b.bc, b.ap, blk); err != nil {
-			return moved, bCheckinLower
-		}
-
-		b.confirmedHeight = h
-		b.startHeight = h + 1
-		l.Info().Uint64("confirmedHeight", b.confirmedHeight).Msg("replace old dummy.")
-		return true, bCheckinValid
-	}
-	if b.blocks[h] != nil {
+	if blk, ok := b.blocks[h]; ok && blk != nil {
 		return moved, bCheckinExisting
 	}
-	if b.startHeight+b.size <= h {
+	if h > b.confirmedHeight+b.size {
 		return moved, bCheckinHigher
 	}
 	b.blocks[h] = blk
-
-	for syncHeight := b.startHeight; syncHeight < b.startHeight+b.size; syncHeight++ {
-		if b.blocks[syncHeight] == nil {
-			continue
-		}
-		if err := commitBlock(b.bc, b.ap, b.blocks[syncHeight]); err == nil {
-			syncedHeight = syncHeight
-			b.confirmedHeight = syncedHeight
-			l.Info().Uint64("syncedHeight", syncedHeight).Msg("Successfully committed block.")
-			delete(b.blocks, syncHeight)
-			continue
-		} else {
-			l.Error().Err(err).Msg("Failed to commit the block")
+	l := logger.With().
+		Uint64("recvHeight", blk.Height()).
+		Uint64("confirmedHeight", b.confirmedHeight).
+		Str("source", "blockBuffer").
+		Logger()
+	for b.size > 0 {
+		next := b.confirmedHeight + 1
+		blk, ok := b.blocks[next]
+		if !ok || blk == nil {
+			break
```

Comment:
yes, break is better than continue 👍  if we don't have tip+1, no need to check further